src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala

```diff
@@ -102,7 +102,13 @@ case class WriteIntoDelta(
       deltaLog.fs.mkdirs(deltaLog.logPath)
     }
 
-    val newFiles = txn.writeFiles(data, Some(options))
+    // If data has no rows, we don't need to write any new files.
+    val newFiles = if (data.take(1).isEmpty) {
+      Seq.empty
+    } else {
+      txn.writeFiles(data, Some(options))
```

Comment:
is it not possible to filter out the zero-byte files from `newFiles`? the reason i ask is that `data.take(1)` has unexpected performance problems in how it is implemented. If there are a lot of empty partitions, it can incrementally keep running multiple Spark jobs scanning through the partitions in an attempt to find that 1 row. I would rather not run a Spark job here just find whether it is empty. Rather I would investigate whether it is possible to identify and eliminate empty files by directly looking at the stats in AddFile.

Reply:
Updated the PR with a new implementation that uses a new StatsTracker to determine which Parquet files contain no data rows.